TulparAudio/internal/include/tulpar/internal/BufferCollection.hpp

```diff
@@ -0,0 +1,167 @@
+{
+public:
+    //! Shortcut to buffer handle type
+    using Handle = audio::Buffer::Handle;
+
+    //! Shortcut to buffer migration mapping as hashmap<old, new>
+    using MigrationMapping = std::unordered_map<audio::Buffer::Handle, audio::Buffer::Handle>;
+
+    /** @brief  Constructs buffer collection object
+     *
+     *  @param  generator   functor that will be called when generating
+     *                      new handle
+     *  @param  reclaimer   functor that will be called when reclaiming
+     *                      previously used handle
+     *  @param  deleter     functor that will be called when releasing
+     *                      previously used handle
+     */
+    BufferCollection(
+        Collection<audio::Buffer>::HandleGenerator generator = OpenAVBufferHandler::Generate
+        , Collection<audio::Buffer>::HandleReclaimer reclaimer = OpenAVBufferHandler::Reclaim
+        , Collection<audio::Buffer>::HandleDeleter deleter = OpenAVBufferHandler::Delete
+    );
+
+    /** @brief  Destructs buffer collection */
+    virtual ~BufferCollection();
+
+    /** @brief  Migrates buffers from given collection
+     *
+     *  @note   Context::MakeCurrent() has to be called on new context prior to
+     *          calling this method in order to migrate buffers
+     *
+     *  @attention  all Buffer objects from @p other are updated to point to
+     *              this collection with new handles
+     *
+     *  @param  other   collection to be copied
+     *
+     *  @return buffer handle mapping
+     */
+    MigrationMapping InheritCollection(BufferCollection const& other);
+
+    //! Returns buffer name
+    std::string GetBufferName(Handle handle) const;
+
+    //! Sets buffer name
+    void SetBufferName(Handle handle, std::string const& name);
+
+    //! Returns buffer channel count
+    uint8_t GetBufferChannelCount(Handle handle) const;
+
+    //! Returns buffer frequency
+    uint32_t GetBufferFrequencyHz(Handle handle) const;
+
+    //! Returns buffer sample count
+    uint32_t GetBufferSampleCount(Handle handle) const;
+
+    //! Returns buffer duration
+    std::chrono::nanoseconds GetBufferDuration(Handle handle) const;
+
+    /** @brief  Initializes buffer with given data
+     *
+     *  Apart from initializing buffer with audio data, parses and sets
+     *  metadata, see @ref BufferInfo for more details.
+     *
+     *  @param  handle  valid buffer handle
+     *  @param  asset   asset handle to audio content
+     *
+     *  @return @c true if data was set successfully, @c false otherwise
+     */
+    bool SetBufferData(Handle handle, mule::asset::Handler asset);
+
+    //! Resets given buffer
+    void ResetBuffer(Handle handle);
+
+protected:
+    //! Creates a buffer object associated with this collection and given handle
+    virtual audio::Buffer CreateObject(Handle handle) override final;
+
+private:
+    //! Meta information for initialized buffer handles
+    struct BufferInfo
```

Comment:
Maybe better `BufferDescriptor`, as is the case with `Device`? Since both are meta info and it would make sense for them to have a similar naming convention.

Reply:
These 2 classes have different purposes and different user-visibility so I don't feel like consistent naming scheme would apply:
* `TulparConfigurator::Device` is a user-visible class describing device entry for configuration purposes.
* `BufferCollection::BufferInfo` is a private meta-information class that's used internally to cache buffer parameters.

I'm not sure if naming it `BufferDescriptor` improves anything.